ethereum/core/src/main/java/tech/pegasys/teku/core/StateGenerator.java

```diff
@@ -13,64 +13,137 @@
    */
-  public Map<Bytes32, BeaconState> produceStatesForBlocks(
-      final Bytes32 baseStateBlockRoot,
-      final BeaconState baseState,
-      final Collection<SignedBeaconBlock> newBlocks) {
-    final Map<Bytes32, BeaconState> statesByRoot = new HashMap<>();
-
-    // Initialize states with the base state
-    statesByRoot.put(baseStateBlockRoot, baseState);
-
-    // Index blocks by parent root
-    final Map<Bytes32, List<SignedBeaconBlock>> blocksByParent = new HashMap<>();
-    for (SignedBeaconBlock currentBlock : newBlocks) {
-      final List<SignedBeaconBlock> blockList =
-          blocksByParent.computeIfAbsent(currentBlock.getParent_root(), (key) -> new ArrayList<>());
-      blockList.add(currentBlock);
+  public BeaconState regenerateStateForBlock(final Bytes32 blockRoot) {
+    return regenerateStateForBlock(blockRoot, Collections.emptyMap());
+  }
+
+  private BeaconState regenerateStateForBlock(
+      final Bytes32 blockRoot, final Map<Bytes32, BeaconState> stateCache) {
+    final Bytes32 rootBlockHash = blockTree.getRootBlock().getRoot();
+    if (rootBlockHash.equals(blockRoot)) {
+      return rootState;
     }
 
-    // Generate states
-    final Deque<Bytes32> parentRoots = new ArrayDeque<>();
-    parentRoots.push(baseStateBlockRoot);
-    while (!parentRoots.isEmpty()) {
-      final Bytes32 parentRoot = parentRoots.pop();
-      final BeaconState parentState = statesByRoot.get(parentRoot);
-      final List<SignedBeaconBlock> blocks =
-          blocksByParent.computeIfAbsent(parentRoot, (key) -> Collections.emptyList());
-      for (SignedBeaconBlock block : blocks) {
-        final Bytes32 blockRoot = block.getMessage().hash_tree_root();
-        final BeaconState state = processBlock(parentState, block);
-        statesByRoot.put(blockRoot, state);
-        parentRoots.push(blockRoot);
+    // Walk from target block towards root of the tree, stopping when we find an available state
+    final List<SignedBeaconBlock> blocks = new ArrayList<>();
+    Optional<SignedBeaconBlock> block = blockTree.getBlock(blockRoot);
+    BeaconState baseState = null;
+    while (block.isPresent()) {
+      final Bytes32 root = block.get().getRoot();
+      baseState = root.equals(rootBlockHash) ? rootState : stateCache.get(block.get().getRoot());
+      if (baseState != null) {
+        break;
       }
+      blocks.add(block.get());
+      block = blockTree.getBlock(block.get().getParent_root());
+    }
+    checkArgument(
+        blocks.size() > 0,
+        "Block %s does not belong to this %s.",
+        blockRoot,
+        getClass().getSimpleName());
+
+    // Process blocks in order
+    Collections.reverse(blocks);
+    BeaconState prevState = baseState;
+    SignedBeaconBlock currentBlock = null;
+    BeaconState currentState = null;
+    for (int i = 0; i < blocks.size(); i++) {
+      currentBlock = blocks.get(i);
+      currentState = processBlock(prevState, currentBlock);
+      prevState = currentState;
```

Comment:
It's a bit confusing that after this loop we wind up using `prevState` for the resulting state after applying the last block.  Would probably be simpler to just remove `prevState` and use only currentState so this winds up as:
```suggestion
      currentState = processBlock(currentState, currentBlock);
```

Reply:
Possibly just `state` rather than `currentState` even.